### PR TITLE
[fix]: [PIE-6784]: Fixing matrix in ci for complex objects and going with the expressionEvaluater route instead of string replace

### PR DESCRIPTION
**Pull request by brijeshdhakar**
## Describe your changes

Using the expressionEvaluator and functor to resolve the matrix expressions in CI Initialise step instead of string replace. String replace was not working properly when complex objects were passed as matrix config. 

Testing: Did basic testing that the matrix expressions are resolving properly.

ReleaseNotes summary needed? No

Is it new task type? No
Did you make changes to the code in a backward-compatible manner? Yes
Did you add a new class which is added in kryo serialization? - No
Will the new version of the delegate receive old version task and process? Yes
If new version of the pipeline-service receives response for the old task will it be able to handle it successfully? Yes
Did you make a change in delegate task which will be created in pipeline-services via grpc call to manager? No change
If the older version of pipeline-service creates the task parameter and calls newer cg manager, will it be okay? Yes
if the newer version of pipeline-service creates the task parameter and calls older cg manager, will it be okay? Yes
Did you remove an enum → No
Did you add an enum(That too at end of list) → No
FeatureFlag Added/Removed -> No
Verify Changes with affects any -> No

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/41204)

### CODE REVIEW

File: clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/strategy/StrategyUtilFunctor.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+public class StrategyUtilFunctor implements LateBindingValue {
+  private Map<String, String> combinations;
+  private int currentIteration;
+  private int totalIteration;
+  private String itemValue;
+  @Override
+  public Object bind() {
+    Map<String, Object> strategyObjectMap = new HashMap<>();
+    Map<String, Object> matrixObjectMap = new HashMap<>();
+    for (Map.Entry<String, String> entry : combinations.entrySet()) {
```
Comment (sahilhindwani-harness) [inline clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/strategy/StrategyUtilFunctor.java:27]: Can we use the same util as we are using in normal StrategyFunctor? 
Maybe we can share some common code?

**Comment by sahilhindwani-harness** [comment]
Add tests for the given scenerio in StrategyUtilsTest

**Review by sahilhindwani-harness — CHANGES_REQUESTED** [review]
> Reviewed 1 of 4 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41204)* status: 5 file(s) to review. 1 change requests. 0 of 2 approvals obtained.

**Review by brijeshdhakar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41204)* status: 7 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/strategy/StrategyUtilFunctor.java` line 27 at r2](https://reviewable.io/reviews/harness/harness-core/41204#-NIcJzbugFT2XCnNjCQP-r2-27:-NIkiusR51WuSj4dJqzp:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/721b97337ea2217e91a05de0282c6697e65587f7/clients/pipeline-service/pms-sdk-core/src/main/java/io/harness/plancreator/strategy/StrategyUtilFunctor.java#L27)):*
<details><summary><i>Previously, sahilhindwani-harness (Sahil Hindwani) wrote…</i></summary><blockquote>

Can we use the same util as we are using in normal StrategyFunctor? 
Maybe we can share some common code?
</blockquote></details>

Done.

**Review by sahilhindwani-harness — APPROVED** [review]
> Reviewed 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41204)* status: 7 file(s) to review. 1 of 2 approvals obtained.

**Review by prashantsharma04 — APPROVED** [review]
> Reviewed 1 of 4 files at r1, 2 of 4 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/41204)* status: :shipit: complete! ✓

**Pull request merged**